<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.555">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>glycowork - ml</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 50,
  "keyboard-shortcut": [
    "f",
    "/",
    "s"
  ],
  "show-item-context": false,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-text-placeholder": "",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


<link rel="stylesheet" href="styles.css">
<meta property="og:title" content="glycowork - ml">
<meta property="og:description" content="Package for processing and analyzing glycans">
<meta property="og:site_name" content="glycowork">
<meta name="twitter:title" content="glycowork - ml">
<meta name="twitter:description" content="Package for processing and analyzing glycans">
<meta name="twitter:card" content="summary">
</head>

<body class="nav-sidebar floating nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg " data-bs-theme="dark">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container mx-auto">
    <a class="navbar-brand" href="./index.html">
    <span class="navbar-title">glycowork</span>
    </a>
  </div>
        <div class="quarto-navbar-tools tools-end">
</div>
          <div id="quarto-search" class="" title="Search"></div>
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
        <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./ml.html">ml</a></li></ol></nav>
        <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
        </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal quarto-sidebar-collapse-item sidebar-navigation floating overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">glycowork</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./core.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Overview</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./glycan_data.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">glycan_data</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./ml.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">ml</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./motif.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">motif</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./network.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">network</span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./examples.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Examples</span></a>
  </div>
</li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" class="quarto-sidebar-collapse-item" data-bs-toggle="collapse" data-bs-target=".quarto-sidebar-collapse-item"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#model_training" id="toc-model_training" class="nav-link active" data-scroll-target="#model_training">model_training</a>
  <ul class="collapse">
  <li><a href="#earlystopping" id="toc-earlystopping" class="nav-link" data-scroll-target="#earlystopping">EarlyStopping</a></li>
  <li><a href="#train_model" id="toc-train_model" class="nav-link" data-scroll-target="#train_model">train_model</a></li>
  <li><a href="#training_setup" id="toc-training_setup" class="nav-link" data-scroll-target="#training_setup">training_setup</a></li>
  <li><a href="#train_ml_model" id="toc-train_ml_model" class="nav-link" data-scroll-target="#train_ml_model">train_ml_model</a></li>
  <li><a href="#analyze_ml_model" id="toc-analyze_ml_model" class="nav-link" data-scroll-target="#analyze_ml_model">analyze_ml_model</a></li>
  <li><a href="#get_mismatch" id="toc-get_mismatch" class="nav-link" data-scroll-target="#get_mismatch">get_mismatch</a></li>
  </ul></li>
  <li><a href="#models" id="toc-models" class="nav-link" data-scroll-target="#models">models</a>
  <ul class="collapse">
  <li><a href="#sweetnet" id="toc-sweetnet" class="nav-link" data-scroll-target="#sweetnet">SweetNet</a></li>
  <li><a href="#lectinoracle" id="toc-lectinoracle" class="nav-link" data-scroll-target="#lectinoracle">LectinOracle</a></li>
  <li><a href="#lectinoracle_flex" id="toc-lectinoracle_flex" class="nav-link" data-scroll-target="#lectinoracle_flex">LectinOracle_flex</a></li>
  <li><a href="#nsequonpred" id="toc-nsequonpred" class="nav-link" data-scroll-target="#nsequonpred">NSequonPred</a></li>
  <li><a href="#init_weights" id="toc-init_weights" class="nav-link" data-scroll-target="#init_weights">init_weights</a></li>
  <li><a href="#prep_model" id="toc-prep_model" class="nav-link" data-scroll-target="#prep_model">prep_model</a></li>
  </ul></li>
  <li><a href="#processing" id="toc-processing" class="nav-link" data-scroll-target="#processing">processing</a>
  <ul class="collapse">
  <li><a href="#dataset_to_graphs" id="toc-dataset_to_graphs" class="nav-link" data-scroll-target="#dataset_to_graphs">dataset_to_graphs</a></li>
  <li><a href="#dataset_to_dataloader" id="toc-dataset_to_dataloader" class="nav-link" data-scroll-target="#dataset_to_dataloader">dataset_to_dataloader</a></li>
  <li><a href="#split_data_to_train" id="toc-split_data_to_train" class="nav-link" data-scroll-target="#split_data_to_train">split_data_to_train</a></li>
  </ul></li>
  <li><a href="#inference" id="toc-inference" class="nav-link" data-scroll-target="#inference">inference</a>
  <ul class="collapse">
  <li><a href="#glycans_to_emb" id="toc-glycans_to_emb" class="nav-link" data-scroll-target="#glycans_to_emb">glycans_to_emb</a></li>
  <li><a href="#get_lectin_preds" id="toc-get_lectin_preds" class="nav-link" data-scroll-target="#get_lectin_preds">get_lectin_preds</a></li>
  <li><a href="#get_nsequon_preds" id="toc-get_nsequon_preds" class="nav-link" data-scroll-target="#get_nsequon_preds">get_Nsequon_preds</a></li>
  <li><a href="#get_esm1b_representations" id="toc-get_esm1b_representations" class="nav-link" data-scroll-target="#get_esm1b_representations">get_esm1b_representations</a></li>
  </ul></li>
  <li><a href="#train_test_split" id="toc-train_test_split" class="nav-link" data-scroll-target="#train_test_split">train_test_split</a>
  <ul class="collapse">
  <li><a href="#hierarchy_filter" id="toc-hierarchy_filter" class="nav-link" data-scroll-target="#hierarchy_filter">hierarchy_filter</a></li>
  <li><a href="#general_split" id="toc-general_split" class="nav-link" data-scroll-target="#general_split">general_split</a></li>
  <li><a href="#prepare_multilabel" id="toc-prepare_multilabel" class="nav-link" data-scroll-target="#prepare_multilabel">prepare_multilabel</a></li>
  </ul></li>
  </ul>
<div class="toc-actions"><ul><li><a href="https://github.com/BojarLab/glycowork/tree/master/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">ml</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->
<p><code>ml</code> contains the code base to process glycan for machine learning, construct state-of-the-art machine learning models, train them, and analyze trained models + glycan representations. It currently contains the following modules:</p>
<ul>
<li><code>model_training</code> contains functions for training machine learning models</li>
<li><code>models</code> describes some examples for machine learning architectures applicable to glycans</li>
<li><code>processing</code> contains helper functions to prepare glycan data for model training</li>
<li><code>inference</code> can be used to analyze trained models, make predictions, or obtain glycan representations</li>
<li><code>train_test_split</code> contains various data split functions to get appropriate training and test sets</li>
</ul>
<section id="model_training" class="level2">
<h2 class="anchored" data-anchor-id="model_training">model_training</h2>
<blockquote class="blockquote">
<p>contains functions for training machine learning models</p>
</blockquote>
<hr>
<section id="earlystopping" class="level3">
<h3 class="anchored" data-anchor-id="earlystopping">EarlyStopping</h3>
<blockquote class="blockquote">
<pre><code> EarlyStopping (patience=7, verbose=False)</code></pre>
</blockquote>
<p><em>Early stops the training if validation loss doesn’t improve after a given patience.</em></p>
<hr>
</section>
<section id="train_model" class="level3">
<h3 class="anchored" data-anchor-id="train_model">train_model</h3>
<blockquote class="blockquote">
<pre><code> train_model (model, dataloaders, criterion, optimizer, scheduler,
              num_epochs=25, patience=50, mode='classification',
              mode2='multi')</code></pre>
</blockquote>
<p>*trains a deep learning model on predicting glycan properties</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="even">
<td style="text-align: left;">dataloaders (PyTorch object): dictionary of dataloader objects with keys ‘train’ and ‘val’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">criterion (PyTorch object): PyTorch loss function</td>
</tr>
<tr class="even">
<td style="text-align: left;">optimizer (PyTorch object): PyTorch optimizer</td>
</tr>
<tr class="odd">
<td style="text-align: left;">scheduler (PyTorch object): PyTorch learning rate decay</td>
</tr>
<tr class="even">
<td style="text-align: left;">num_epochs (int): number of epochs for training; default:25</td>
</tr>
<tr class="odd">
<td style="text-align: left;">patience (int): number of epochs without improvement until early stop; default:50</td>
</tr>
<tr class="even">
<td style="text-align: left;">mode (string): ‘classification’, ‘multilabel’, or ‘regression’; default:classification</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mode2 (string): further specifying classification into ‘multi’ or ‘binary’ classification;default:multi</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns the best model seen during training*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="training_setup" class="level3">
<h3 class="anchored" data-anchor-id="training_setup">training_setup</h3>
<blockquote class="blockquote">
<pre><code> training_setup (model, lr, lr_patience=4, factor=0.2,
                 weight_decay=0.0001, mode='multiclass', num_classes=2,
                 gsam_alpha=0.0)</code></pre>
</blockquote>
<p>*prepares optimizer, learning rate scheduler, and loss criterion for model training</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (PyTorch object): graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="even">
<td style="text-align: left;">lr (float): learning rate</td>
</tr>
<tr class="odd">
<td style="text-align: left;">lr_patience (int): number of epochs without validation loss improvement before reducing the learning rate;default:4</td>
</tr>
<tr class="even">
<td style="text-align: left;">factor (float): factor by which learning rate is multiplied upon reduction</td>
</tr>
<tr class="odd">
<td style="text-align: left;">weight_decay (float): regularization parameter for the optimizer; default:0.001</td>
</tr>
<tr class="even">
<td style="text-align: left;">mode (string): ‘multiclass’: classification with multiple classes, ‘multilabel’: predicting several labels at the same time, ‘binary’:binary classification, ‘regression’: regression; default:‘multiclass’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">num_classes (int): number of classes; only used when mode == ‘multiclass’ or ‘multilabel’</td>
</tr>
<tr class="even">
<td style="text-align: left;">gsam_alpha (float): if higher than zero, uses GSAM instead of SAM for the optimizer</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns optimizer, learning rate scheduler, and loss criterion objects*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="train_ml_model" class="level3">
<h3 class="anchored" data-anchor-id="train_ml_model">train_ml_model</h3>
<blockquote class="blockquote">
<pre><code> train_ml_model (X_train, X_test, y_train, y_test, mode='classification',
                 feature_calc=False, return_features=False,
                 feature_set=['known', 'exhaustive'],
                 additional_features_train=None,
                 additional_features_test=None)</code></pre>
</blockquote>
<p>*wrapper function to train standard machine learning models on glycans</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">X_train, X_test (list or dataframe): either lists of glycans (needs feature_calc = True) or motif dataframes such as from annotate_dataset</td>
</tr>
<tr class="even">
<td style="text-align: left;">y_train, y_test (list): lists of labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">mode (string): ‘classification’ or ‘regression’; default:‘classification’</td>
</tr>
<tr class="even">
<td style="text-align: left;">feature_calc (bool): set to True for calculating motifs from glycans; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">return_features (bool): whether to return calculated features; default:False</td>
</tr>
<tr class="even">
<td style="text-align: left;">feature_set (list): which feature set to use for annotations, add more to list to expand; default:[‘known’,‘exhaustive’]; options are: ‘known’ (hand-crafted glycan features), ‘graph’ (structural graph features of glycans), and ‘exhaustive’ (all mono- and disaccharide features)</td>
</tr>
<tr class="odd">
<td style="text-align: left;">additional_features_train (dataframe): additional features (apart from glycans) to be used for training. Has to be of the same length as X_train; default:None</td>
</tr>
<tr class="even">
<td style="text-align: left;">additional_features_test (dataframe): additional features (apart from glycans) to be used for evaluation. Has to be of the same length as X_test; default:None</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns trained model*</td>
</tr>
</tbody>
</table>
<div id="separate-knock" class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>human <span class="op">=</span> [<span class="dv">1</span> <span class="cf">if</span> k <span class="op">==</span> <span class="st">'Homo_sapiens'</span> <span class="cf">else</span> <span class="dv">0</span> <span class="cf">for</span> k <span class="kw">in</span> df_species[df_species.Order<span class="op">==</span><span class="st">'Primates'</span>].Species.values.tolist()]</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>X_train, X_test, y_train, y_test <span class="op">=</span> general_split(df_species[df_species.Order<span class="op">==</span><span class="st">'Primates'</span>].glycan.values.tolist(), human)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>model_ft, _, X_test <span class="op">=</span> train_ml_model(X_train, X_test, y_train, y_test, feature_calc <span class="op">=</span> <span class="va">True</span>, feature_set <span class="op">=</span> [<span class="st">'terminal'</span>],</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>                         return_features <span class="op">=</span> <span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Calculating Glycan Features...

Training model...

Evaluating model...
Accuracy of trained model on separate validation set: 0.8302047781569966</code></pre>
</div>
</div>
<hr>
</section>
<section id="analyze_ml_model" class="level3">
<h3 class="anchored" data-anchor-id="analyze_ml_model">analyze_ml_model</h3>
<blockquote class="blockquote">
<pre><code> analyze_ml_model (model)</code></pre>
</blockquote>
<p>*plots relevant features for model prediction</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (model object): trained machine learning model from train_ml_model*</td>
</tr>
</tbody>
</table>
<div id="herbal-holmes" class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>analyze_ml_model(model_ft)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="02_ml_files/figure-html/cell-8-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<hr>
</section>
<section id="get_mismatch" class="level3">
<h3 class="anchored" data-anchor-id="get_mismatch">get_mismatch</h3>
<blockquote class="blockquote">
<pre><code> get_mismatch (model, X_test, y_test, n=10)</code></pre>
</blockquote>
<p>*analyzes misclassifications of trained machine learning model</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (model object): trained machine learning model from train_ml_model</td>
</tr>
<tr class="even">
<td style="text-align: left;">X_test (dataframe): motif dataframe used for validating model</td>
</tr>
<tr class="odd">
<td style="text-align: left;">y_test (list): list of labels</td>
</tr>
<tr class="even">
<td style="text-align: left;">n (int): number of returned misclassifications; default:10</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns tuples of misclassifications and their predicted probability*</td>
</tr>
</tbody>
</table>
<div id="incredible-basket" class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>get_mismatch(model_ft, X_test, y_test)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[('GlcNAc6S(b1-3)Gal(b1-4)Glc-ol', 0.8178003430366516),
 ('Fuc(a1-?)[Gal(b1-?)]GlcNAc(b1-2)[Fuc(a1-?)[Gal(b1-?)]GlcNAc(b1-4)]Man(a1-3)[Fuc(a1-?)[Gal(b1-?)]GlcNAc(b1-?)[GlcNAc(b1-?)]Man(a1-6)][GlcNAc(b1-4)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',
  0.4119275212287903),
 ('Neu5Ac(a2-6)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-?)[Gal(b1-?)GlcNAc(b1-?)]Man(a1-3)[Gal(b1-?)GlcNAc(b1-2)[Gal(b1-?)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.6590214967727661),
 ('Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.7636852860450745),
 ('Man(a1-?)Man(a1-?)[Man(a1-?)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',
  0.7674980759620667),
 ('Neu5Gc(a2-3)Gal(b1-4)Glc-ol', 0.5243940949440002),
 ('Gal(b1-3)[Neu5Ac(a2-6)]GlcNAc(b1-3)Gal(b1-4)Glc-ol', 0.7285189628601074),
 ('Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-2)Man(a1-6)[Man(a1-3)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc',
  0.7636852860450745),
 ('Neu5Ac(a2-3)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-2)[Neu5Ac(a2-6)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-4)]Man(a1-3)[Neu5Ac(a2-6)Gal(b1-?)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',
  0.5886961221694946),
 ('Neu5Ac(a2-3)Gal(b1-?)[Fuc(a1-?)]GlcNAc(b1-?)[GlcNAc(b1-?)]Man(a1-3)[Gal(b1-?)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc',
  0.44415077567100525)]</code></pre>
</div>
</div>
</section>
</section>
<section id="models" class="level2">
<h2 class="anchored" data-anchor-id="models">models</h2>
<blockquote class="blockquote">
<p>describes some examples for machine learning architectures applicable to glycans. The main portal is prep_models which allows users to setup (trained) models by their string names</p>
</blockquote>
<hr>
<section id="sweetnet" class="level3">
<h3 class="anchored" data-anchor-id="sweetnet">SweetNet</h3>
<blockquote class="blockquote">
<pre><code> SweetNet (lib_size, num_classes=1)</code></pre>
</blockquote>
<p>*given glycan graphs as input, predicts properties via a graph neural network</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">lib_size (int): number of unique tokens for graph nodes; usually len(lib)</td>
</tr>
<tr class="even">
<td style="text-align: left;">num_classes (int): number of output classes; only &gt;1 for multilabel classification; default:1</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns batch-wise predictions*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="lectinoracle" class="level3">
<h3 class="anchored" data-anchor-id="lectinoracle">LectinOracle</h3>
<blockquote class="blockquote">
<pre><code> LectinOracle (input_size_glyco, hidden_size=128, num_classes=1,
               data_min=-11.355, data_max=23.892, input_size_prot=1280)</code></pre>
</blockquote>
<p>*given glycan graphs and protein representations as input, predicts protein-glycan binding</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">input_size_glyco (int): number of unique tokens for graph nodes; usually len(lib)</td>
</tr>
<tr class="even">
<td style="text-align: left;">hidden_size (int): layer size for the graph convolutions; default:128</td>
</tr>
<tr class="odd">
<td style="text-align: left;">num_classes (int): number of output classes; only &gt;1 for multilabel classification; default:1</td>
</tr>
<tr class="even">
<td style="text-align: left;">data_min (float): minimum observed value in training data; default: -11.355</td>
</tr>
<tr class="odd">
<td style="text-align: left;">data_max (float): maximum observed value in training data; default: 23.892</td>
</tr>
<tr class="even">
<td style="text-align: left;">input_size_prot (int): dimensionality of protein representations used as input; default:1280</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns batch-wise predictions*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="lectinoracle_flex" class="level3">
<h3 class="anchored" data-anchor-id="lectinoracle_flex">LectinOracle_flex</h3>
<blockquote class="blockquote">
<pre><code> LectinOracle_flex (input_size_glyco, hidden_size=128, num_classes=1,
                    data_min=-11.355, data_max=23.892,
                    input_size_prot=1000)</code></pre>
</blockquote>
<p>*given glycan graphs and protein sequences as input, predicts protein-glycan binding</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">input_size_glyco (int): number of unique tokens for graph nodes; usually len(lib)</td>
</tr>
<tr class="even">
<td style="text-align: left;">hidden_size (int): layer size for the graph convolutions; default:128</td>
</tr>
<tr class="odd">
<td style="text-align: left;">num_classes (int): number of output classes; only &gt;1 for multilabel classification; default:1</td>
</tr>
<tr class="even">
<td style="text-align: left;">data_min (float): minimum observed value in training data; default: -11.355</td>
</tr>
<tr class="odd">
<td style="text-align: left;">data_max (float): maximum observed value in training data; default: 23.892</td>
</tr>
<tr class="even">
<td style="text-align: left;">input_size_prot (int): maximum length of protein sequence for padding/cutting; default:1000</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns batch-wise predictions*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="nsequonpred" class="level3">
<h3 class="anchored" data-anchor-id="nsequonpred">NSequonPred</h3>
<blockquote class="blockquote">
<pre><code> NSequonPred ()</code></pre>
</blockquote>
<p>*given an ESM1b representation of N and 20 AA up + downstream, predicts whether it’s a sequon</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns batch-wise predictions*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="init_weights" class="level3">
<h3 class="anchored" data-anchor-id="init_weights">init_weights</h3>
<blockquote class="blockquote">
<pre><code> init_weights (model, mode='sparse', sparsity=0.1)</code></pre>
</blockquote>
<p>*initializes linear layers of PyTorch model with a weight initialization</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model (Pytorch object): neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="even">
<td style="text-align: left;">mode (string): which initialization algorithm; choices are ‘sparse’,‘kaiming’,‘xavier’;default:‘sparse’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sparsity (float): proportion of sparsity after initialization; default:0.1 / 10%*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="prep_model" class="level3">
<h3 class="anchored" data-anchor-id="prep_model">prep_model</h3>
<blockquote class="blockquote">
<pre><code> prep_model (model_type, num_classes, libr=None, trained=False)</code></pre>
</blockquote>
<p>*wrapper to instantiate model, initialize it, and put it on the GPU</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">model_type (string): string indicating the type of model</td>
</tr>
<tr class="even">
<td style="text-align: left;">num_classes (int): number of unique classes for classification</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns PyTorch model object*</td>
</tr>
</tbody>
</table>
</section>
</section>
<section id="processing" class="level2">
<h2 class="anchored" data-anchor-id="processing">processing</h2>
<blockquote class="blockquote">
<p>contains helper functions to prepare glycan data for model training</p>
</blockquote>
<hr>
<section id="dataset_to_graphs" class="level3">
<h3 class="anchored" data-anchor-id="dataset_to_graphs">dataset_to_graphs</h3>
<blockquote class="blockquote">
<pre><code> dataset_to_graphs (glycan_list, labels, libr=None,
                    label_type=torch.int64)</code></pre>
</blockquote>
<p>*wrapper function to convert a whole list of glycans into a graph dataset</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycan_list (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">labels (list): list of labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
<tr class="even">
<td style="text-align: left;">label_type (torch object): which tensor type for label, default is torch.long for binary labels, change to torch.float for continuous</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns list of node list / edge list / label list data tuples*</td>
</tr>
</tbody>
</table>
<div id="quiet-butter" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>dataset_to_graphs([<span class="st">"Neu5Ac(a2-3)Gal(b1-4)Glc"</span>,</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>], [<span class="dv">1</span>, <span class="dv">0</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>[Data(edge_index=[2, 8], labels=[5], string_labels=[5], num_nodes=5, y=1),
 Data(edge_index=[2, 8], labels=[5], string_labels=[5], num_nodes=5, y=0)]</code></pre>
</div>
</div>
<hr>
</section>
<section id="dataset_to_dataloader" class="level3">
<h3 class="anchored" data-anchor-id="dataset_to_dataloader">dataset_to_dataloader</h3>
<blockquote class="blockquote">
<pre><code> dataset_to_dataloader (glycan_list, labels, libr=None, batch_size=32,
                        shuffle=True, drop_last=False, extra_feature=None,
                        label_type=torch.int64)</code></pre>
</blockquote>
<p>*wrapper function to convert glycans and labels to a torch_geometric DataLoader</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycan_list (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">labels (list): list of labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
<tr class="even">
<td style="text-align: left;">batch_size (int): how many samples should be in each batch; default:32</td>
</tr>
<tr class="odd">
<td style="text-align: left;">shuffle (bool): if samples should be shuffled when making dataloader; default:True</td>
</tr>
<tr class="even">
<td style="text-align: left;">drop_last (bool): whether last batch is dropped; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">extra_feature (list): can be used to feed another input to the dataloader; default:None</td>
</tr>
<tr class="even">
<td style="text-align: left;">label_type (torch object): which tensor type for label, default is torch.long for binary labels, change to torch.float for continuous</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns a dataloader object used for training deep learning models*</td>
</tr>
</tbody>
</table>
<div id="filled-glass" class="cell">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="bu">next</span>(<span class="bu">iter</span>(dataset_to_dataloader([<span class="st">"Neu5Ac(a2-3)Gal(b1-4)Glc"</span>,</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>], [<span class="dv">1</span>, <span class="dv">0</span>])))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>DataBatch(edge_index=[2, 16], labels=[10], string_labels=[2], num_nodes=10, y=[2], batch=[10], ptr=[3])</code></pre>
</div>
</div>
<hr>
</section>
<section id="split_data_to_train" class="level3">
<h3 class="anchored" data-anchor-id="split_data_to_train">split_data_to_train</h3>
<blockquote class="blockquote">
<pre><code> split_data_to_train (glycan_list_train, glycan_list_val, labels_train,
                      labels_val, libr=None, batch_size=32,
                      drop_last=False, extra_feature_train=None,
                      extra_feature_val=None, label_type=torch.int64)</code></pre>
</blockquote>
<p>*wrapper function to convert split training/test data into dictionary of dataloaders</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycan_list_train (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">glycan_list_val (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="odd">
<td style="text-align: left;">labels_train (list): list of labels</td>
</tr>
<tr class="even">
<td style="text-align: left;">labels_val (list): list of labels</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
<tr class="even">
<td style="text-align: left;">batch_size (int): how many samples should be in each batch; default:32</td>
</tr>
<tr class="odd">
<td style="text-align: left;">drop_last (bool): whether last batch is dropped; default:False</td>
</tr>
<tr class="even">
<td style="text-align: left;">extra_feature_train (list): can be used to feed another input to the dataloader; default:None</td>
</tr>
<tr class="odd">
<td style="text-align: left;">extra_feature_val (list): can be used to feed another input to the dataloader; default:None</td>
</tr>
<tr class="even">
<td style="text-align: left;">label_type (torch object): which tensor type for label, default is torch.long for binary labels, change to torch.float for continuous</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns a dictionary of dataloaders for training and testing deep learning models*</td>
</tr>
</tbody>
</table>
<div id="reduced-female" class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>split_data_to_train([<span class="st">"Neu5Ac(a2-3)Gal(b1-4)Glc"</span>, <span class="st">"Fuc(a1-2)Gal(b1-3)GalNAc"</span>],</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                    [<span class="st">"Neu5Ac(a2-6)Gal(b1-4)Glc"</span>, <span class="st">"Fuc(a1-2)Gal(a1-3)GalNAc"</span>],</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                    [<span class="dv">1</span>, <span class="dv">0</span>], [<span class="dv">0</span>,<span class="dv">1</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<pre><code>{'train': &lt;torch_geometric.loader.dataloader.DataLoader&gt;,
 'val': &lt;torch_geometric.loader.dataloader.DataLoader&gt;}</code></pre>
</div>
</div>
</section>
</section>
<section id="inference" class="level2">
<h2 class="anchored" data-anchor-id="inference">inference</h2>
<p><a class="anchor" id="inference"></a> &gt;can be used to analyze trained models, make predictions, or obtain glycan representations</p>
<hr>
<section id="glycans_to_emb" class="level3">
<h3 class="anchored" data-anchor-id="glycans_to_emb">glycans_to_emb</h3>
<blockquote class="blockquote">
<pre><code> glycans_to_emb (glycans, model, libr=None, batch_size=32, rep=True,
                 class_list=None)</code></pre>
</blockquote>
<p>*Returns a dataframe of learned representations for a list of glycans</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycans (list): list of glycans in IUPAC-condensed as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">model (PyTorch object): trained graph neural network (such as SweetNet) for analyzing glycans</td>
</tr>
<tr class="odd">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
<tr class="even">
<td style="text-align: left;">batch_size (int): change to batch_size used during training; default:32</td>
</tr>
<tr class="odd">
<td style="text-align: left;">rep (bool): True returns representations, False returns actual predicted labels; default is True</td>
</tr>
<tr class="even">
<td style="text-align: left;">class_list (list): list of unique classes to map predictions</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns dataframe of learned representations (columns) for each glycan (rows)*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="get_lectin_preds" class="level3">
<h3 class="anchored" data-anchor-id="get_lectin_preds">get_lectin_preds</h3>
<blockquote class="blockquote">
<pre><code> get_lectin_preds (prot, glycans, model, prot_dic=None,
                   background_correction=False, correction_df=None,
                   batch_size=128, libr=None, sort=True, flex=False)</code></pre>
</blockquote>
<p>*Wrapper that uses LectinOracle-type model for predicting binding of protein to glycans</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">prot (string): protein amino acid sequence</td>
</tr>
<tr class="even">
<td style="text-align: left;">glycans (list): list of glycans in IUPACcondensed</td>
</tr>
<tr class="odd">
<td style="text-align: left;">model (PyTorch object): trained LectinOracle-type model</td>
</tr>
<tr class="even">
<td style="text-align: left;">prot_dic (dictionary): dictionary of type protein sequence:ESM1b representation</td>
</tr>
<tr class="odd">
<td style="text-align: left;">background_correction (bool): whether to correct predictions for background; default:False</td>
</tr>
<tr class="even">
<td style="text-align: left;">correction_df (dataframe): background prediction for (ideally) all provided glycans; default:V4 correction file</td>
</tr>
<tr class="odd">
<td style="text-align: left;">batch_size (int): change to batch_size used during training; default:128</td>
</tr>
<tr class="even">
<td style="text-align: left;">libr (dict): dictionary of form glycoletter:index</td>
</tr>
<tr class="odd">
<td style="text-align: left;">sort (bool): whether to sort prediction results descendingly; default:True</td>
</tr>
<tr class="even">
<td style="text-align: left;">flex (bool): depends on whether you use LectinOracle (False) or LectinOracle_flex (True); default:False</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns dataframe of glycan sequences and predicted binding to prot*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="get_nsequon_preds" class="level3">
<h3 class="anchored" data-anchor-id="get_nsequon_preds">get_Nsequon_preds</h3>
<blockquote class="blockquote">
<pre><code> get_Nsequon_preds (prots, model, prot_dic)</code></pre>
</blockquote>
<p>*Predicts whether an N-sequon will be glycosylated</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">prots (list): list of protein sequences (strings), in the form of 20 AA + N + 20 AA; replace missing sequence with corr. number of ‘z’</td>
</tr>
<tr class="even">
<td style="text-align: left;">model (PyTorch object): trained NSequonPred-type model</td>
</tr>
<tr class="odd">
<td style="text-align: left;">prot_dic (dictionary): dictionary of type protein sequence:ESM1b representation</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns dataframe of protein sequences and predicted likelihood of being an N-sequon*</td>
</tr>
</tbody>
</table>
<hr>
</section>
<section id="get_esm1b_representations" class="level3">
<h3 class="anchored" data-anchor-id="get_esm1b_representations">get_esm1b_representations</h3>
<blockquote class="blockquote">
<pre><code> get_esm1b_representations (prots, model, alphabet)</code></pre>
</blockquote>
<p>*Retrieves ESM1b representations of protein for using them as input for LectinOracle</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">prots (list): list of protein sequences (strings) that should be converted</td>
</tr>
<tr class="even">
<td style="text-align: left;">model (ESM1b object): trained ESM1b model; from running esm.pretrained.esm1b_t33_650M_UR50S()</td>
</tr>
<tr class="odd">
<td style="text-align: left;">alphabet (ESM1b object): used for converting sequences; from running esm.pretrained.esm1b_t33_650M_UR50S()</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns dictionary of the form protein sequence:ESM1b representation*</td>
</tr>
</tbody>
</table>
<p>In order to run <code>get_esm1b_representations</code>, you first have to run this snippet:</p>
<p><code>!pip install fair-esm import esm model, alphabet = esm.pretrained.esm1b_t33_650M_UR50S()</code></p>
</section>
</section>
<section id="train_test_split" class="level2">
<h2 class="anchored" data-anchor-id="train_test_split">train_test_split</h2>
<blockquote class="blockquote">
<p>contains various data split functions to get appropriate training and test sets</p>
</blockquote>
<hr>
<section id="hierarchy_filter" class="level3">
<h3 class="anchored" data-anchor-id="hierarchy_filter">hierarchy_filter</h3>
<blockquote class="blockquote">
<pre><code> hierarchy_filter (df_in, rank='Domain', min_seq=5, wildcard_seed=False,
                   wildcard_list=None, wildcard_name=None, r=0.1,
                   col='glycan')</code></pre>
</blockquote>
<p>*stratified data split in train/test at the taxonomic level, removing duplicate glycans and infrequent classes</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">df_in (dataframe): dataframe of glycan sequences and taxonomic labels</td>
</tr>
<tr class="even">
<td style="text-align: left;">rank (string): which rank should be filtered; default:‘domain’</td>
</tr>
<tr class="odd">
<td style="text-align: left;">min_seq (int): how many glycans need to be present in class to keep it; default:5</td>
</tr>
<tr class="even">
<td style="text-align: left;">wildcard_seed (bool): set to True if you want to seed wildcard glycoletters; default:False</td>
</tr>
<tr class="odd">
<td style="text-align: left;">wildcard_list (list): list which glycoletters a wildcard encompasses</td>
</tr>
<tr class="even">
<td style="text-align: left;">wildcard_name (string): how the wildcard should be named in the IUPAC-condensed nomenclature</td>
</tr>
<tr class="odd">
<td style="text-align: left;">r (float): rate of replacement, default:0.1 or 10%</td>
</tr>
<tr class="even">
<td style="text-align: left;">col (string): column name for glycan sequences; default:glycan</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns train_x, val_x (lists of glycans (strings) after stratified shuffle split)</td>
</tr>
<tr class="even">
<td style="text-align: left;">train_y, val_y (lists of taxonomic labels (mapped integers))</td>
</tr>
<tr class="odd">
<td style="text-align: left;">id_val (taxonomic labels in text form (strings))</td>
</tr>
<tr class="even">
<td style="text-align: left;">class_list (list of unique taxonomic classes (strings))</td>
</tr>
<tr class="odd">
<td style="text-align: left;">class_converter (dictionary to map mapped integers back to text labels)*</td>
</tr>
</tbody>
</table>
<div id="latin-gross" class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>train_x, val_x, train_y, val_y, id_val, class_list, class_converter <span class="op">=</span> hierarchy_filter(df_species,</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>                                                                                       rank <span class="op">=</span> <span class="st">'Kingdom'</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(train_x[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['GlcA(b1-3)GlcNAc(a1-2)Gal', 'Neu5Ac(a2-?)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc', 'Gal2Ac3Ac4Ac6Ac(a1-3)Gal2Ac4Ac6Ac(b1-4)GlcNAc3Ac6Ac(b1-3)[Gal2Ac3Ac4Ac6Ac(a1-3)Gal2Ac4Ac6Ac(b1-4)GlcNAc3Ac6Ac(b1-6)]Gal2Ac4Ac(b1-4)GlcNAc3Ac6Ac(b1-3)Gal2Ac4Ac6Ac(b1-4)Glc2Ac3Ac6Ac', 'Neu5Ac(a2-3)Gal(b1-4)GlcNAc(b1-2)Man(a1-3)[Neu5Ac(a2-3)Gal(b1-4)[Fuc(a1-3)]GlcNAc(b1-2)Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc', 'Neu5Ac(a2-?)Gal(b1-3)[GlcNAc(b1-6)]GalNAc', 'AltNAcA(a1-3)D-FucNAcN(b1-4)AltNAcA(a1-3)D-FucNAcN(b1-3)Glc(b1-3)[Gal(a1-2)Gal(a1-2)]Glc(a1-3)Glc(a1-3)[GlcN(a1-7)LDManHep(a1-7)]LDManHepOP(a1-3)LDManHepOPEtN(a1-5)[Kdo(a2-4)]Kdo', 'Pse5Ac7Ac(b2-6)Gal(b1-6)Gal(b1-3)GalNAc(b1-4)Pse5Ac7Ac', '{Gal(b1-4)GlcNAc(b1-?)}{Gal(b1-4)GlcNAc(b1-?)}{Neu5Ac(a2-?)}{Neu5Ac(a2-?)}{Neu5Ac(a2-?)}{Neu5Ac(a2-?)}Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-4)]Man(a1-3)[Gal(b1-4)GlcNAc(b1-2)[Gal(b1-4)GlcNAc(b1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)[Fuc(a1-6)]GlcNAc', 'Man(a1-3)[Man(a1-6)]Man(a1-6)[Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc', 'Glc(a1-4)Glc(a1-4)[Glc(a1-4)Glc(a1-6)]Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc(a1-4)Glc']</code></pre>
</div>
</div>
<hr>
</section>
<section id="general_split" class="level3">
<h3 class="anchored" data-anchor-id="general_split">general_split</h3>
<blockquote class="blockquote">
<pre><code> general_split (glycans, labels, test_size=0.2)</code></pre>
</blockquote>
<p>*splits glycans and labels into train / test sets</p>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">glycans (list): list of IUPAC-condensed glycan sequences as strings</td>
</tr>
<tr class="even">
<td style="text-align: left;">labels (list): list of labels used for prediction</td>
</tr>
<tr class="odd">
<td style="text-align: left;">test_size (float): % size of test set; default:0.2 / 20%</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Returns X_train, X_test, y_train, y_test*</td>
</tr>
</tbody>
</table>
<div id="statistical-harvest" class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>train_x, val_x, train_y, val_y <span class="op">=</span> general_split(df_species.glycan.values.tolist(),</span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>                                              df_species.Species.values.tolist())</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(train_x[:<span class="dv">10</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>['[Glc(b1-6)Glc(b1-6)]Glc(b1-3)Glc(b1-3)Glc', 'Gal4Pyr6Pyr(b1-3)GlcNAc(a1-4)GlcA(b1-3)QuiNAc4NAc(b1-2)Gal4Pyr', 'Man(a1-2)[Man1P(a1-6)]Man(a1-2)[Man(a1-2)Man(a1-6)]Man(a1-3)[Man(a1-3)[Man(a1-2)[Man1P(a1-6)]Man(a1-6)]Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc', 'Neu5Gc(a2-3)Gal(b1-4)GlcNAc(b1-?)[Gal(b1-4)GlcNAc(b1-?)]Gal(b1-4)Glc-ol', 'Rha(a1-3)[Xyl(b1-2)]Rha(a1-3)[Xyl(b1-4)]Rha(a1-2)Rha', 'Glc(a1-2)Glc(a1-3)Glc(a1-3)Man(a1-2)Man(a1-2)Man(a1-3)[Man(a1-6)]Man(b1-4)GlcNAc(b1-4)GlcNAc', 'GlcNAc(b1-2)Man6P(a1-6)[Man(a1-3)]Man(a1-6)[Man(a1-2)Man(a1-3)]Man(b1-4)GlcNAc(b1-4)GlcNAc', 'IdoA(a1-3)GalNAc6S', 'Gal(b1-4)GlcNAc', 'Man(a1-6)[Man(a1-2)][Galf(b1-4)]Man(a1-6)Man(a1-6)Man']</code></pre>
</div>
</div>
<hr>
</section>
<section id="prepare_multilabel" class="level3">
<h3 class="anchored" data-anchor-id="prepare_multilabel">prepare_multilabel</h3>
<blockquote class="blockquote">
<pre><code> prepare_multilabel (df, rank='Species', glycan_col='glycan')</code></pre>
</blockquote>
<p>*converts a one row per glycan-species/tissue/disease association file to a format of one glycan - all associations</p>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Arguments:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">df (dataframe): dataframe where each row is one glycan - species association</td>
</tr>
<tr class="even">
<td style="text-align: left;">rank (string): which label column should be used; default:Species</td>
</tr>
<tr class="odd">
<td style="text-align: left;">glycan_col (string): column name of where the glycan sequences are stored; default:glycan</td>
</tr>
</tbody>
</table>
<table class="caption-top table">
<colgroup>
<col style="width: 100%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">Returns:</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">(1) list of unique glycans in df</td>
</tr>
<tr class="even">
<td style="text-align: left;">(2) list of lists, where each inner list are all the labels of a glycan*</td>
</tr>
</tbody>
</table>
<div id="popular-bailey" class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>glycans, labels <span class="op">=</span> prepare_multilabel(df_species[df_species.Order <span class="op">==</span> <span class="st">'Carnivora'</span>])</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(glycans[<span class="dv">50</span>])</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(labels[<span class="dv">50</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Gal1Cer3S
[1.0, 0.0, 0.0, 1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]</code></pre>
</div>
</div>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp("https:\/\/BojarLab\.github\.io\/glycowork\/");
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




<footer class="footer"><div class="nav-footer"><div class="nav-footer-center"><div class="toc-actions d-sm-block d-md-none"><ul><li><a href="https://github.com/BojarLab/glycowork/tree/master/issues/new" class="toc-action"><i class="bi bi-github"></i>Report an issue</a></li></ul></div></div></div></footer></body></html>